src/umode.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) 2022 by Rivos Inc.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::smp::PerCpu;
+use riscv_regs::GeneralPurposeRegisters;
+use s_mode_utils::print::*;
+use spin::{Mutex, MutexGuard, Once};
+
+/// Host GPR and which must be saved/restored when entering/exiting a task.
+#[derive(Default)]
+#[repr(C)]
+struct HostCpuRegs {
+    gprs: GeneralPurposeRegisters,
+    stvec: u64,
+    sscratch: u64,
+}
+
+/// Umode GPR and CSR state which must be saved/restored when exiting/entering a task.
+#[derive(Default)]
+#[repr(C)]
+struct UmodeCpuRegs {
+    gprs: GeneralPurposeRegisters,
+    sepc: u64,
+}
+
+/// CSRs written on an exit from virtualization that are used by the host to determine the cause of
+/// the trap.
+#[derive(Default)]
+#[repr(C)]
+struct TrapRegs {
+    scause: u64,
+    stval: u64,
+}
+
+/// CPU register state that must be saved or restored when entering/exiting a task.
+#[derive(Default)]
+#[repr(C)]
+struct UmodeCpuArchState {
+    host_regs: HostCpuRegs,
+    task_regs: UmodeCpuRegs,
+    trap_csrs: TrapRegs,
+}
+
+/// The loaded U-mode task.
+pub static UMODE_TASK: Once<Umode> = Once::new();
+
+/// Errors returned by Umode.
+#[derive(Debug)]
+pub enum Error {}
+
+/// Salus U-mode global data.
+pub struct Umode {
+    entry: u64,
+}
+
+impl Umode {
+    /// Initialize U-mode.
+    pub fn init(entry: u64) {
+        println!("U-mode entry at {:016x}\n", entry);
+        let umode = Umode { entry };
+        UMODE_TASK.call_once(|| umode);
+    }
+
+    /// Create a new umode state. This must be done once on every physical CPU.
+    pub fn new_cpu_state(&self) -> PerCpuUmode {
+        PerCpuUmode {
+            entry: self.entry,
+            arch: Mutex::new(UmodeCpuArchState::default()),
+        }
+    }
+}
+
+/// Per-CPU U-mode state.
+pub struct PerCpuUmode {
+    entry: u64,
+    arch: Mutex<UmodeCpuArchState>,
```

Comment:
As we were talking about yesterday, have a `Mutex` protect a per-CPU struct is kinda pointless :). You can use a `RefCell` here instead to get interior mutability, like we do already for `PerCpu::vmid_tracker`.

Reply:
Agreed. :-) Will do.

Reply:
Changed to `RefCell`.